### PR TITLE
Query result cache to speed up query processing for rarely changing datasets

### PR DESCRIPTION
**Pull request by ghost**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (up to few sentences, required except for Non-significant/Documentation categories):

There are many unchanging data in analytical databases.
In this case, it is possible to get better performance if ClickHouse stores the calculated data as caches and reuse them.
So I am developing a query cache feature.

Detailed description (optional):

## Local Query Cache
Local Query Cache works on IN-operator of subqueries.
It stores the result of subqueries for IN-operator and reuses it later for the same subqueries.
When there is any update on the reference table of query-cache-result, then it invalidates the query-cache-result. But this feature is implemented only for MergeTree engines.
If you want to use this feature, turn on an option `SET use_experimental_local_query_cache = 1`.

**Create a test table**
```SQL
CREATE TABLE test_cache
(
    `dt` Date, 
    `id` int, 
    `name` String
)
ENGINE = MergeTree()
PARTITION BY dt
ORDER BY id;

INSERT INTO test_cache SELECT '2019-12-01', number AS id, CAST(id, 'String') FROM system.numbers LIMIT 10000000;
INSERT INTO test_cache SELECT '2019-12-02', number AS id, CAST(id, 'String') FROM system.numbers LIMIT 10000000;
INSERT INTO test_cache SELECT '2019-12-03', number AS id, CAST(id, 'String') FROM system.numbers LIMIT 10000000;
INSERT INTO test_cache SELECT '2019-12-04', number AS id, CAST(id, 'String') FROM system.numbers LIMIT 10000000;
INSERT INTO test_cache SELECT '2019-12-05', number AS id, CAST(id, 'String') FROM system.numbers LIMIT 10000000;
```

**Query performance**
```SQL
SET use_experimental_local_query_cache = 1

SELECT count()
FROM test_cache
WHERE (dt = '2019-12-05') AND (id IN 
(
    SELECT id
    FROM test_cache
    WHERE (dt >= '2019-12-01') AND (dt <= '2019-12-04')
    GROUP BY id
))
```

First time, this query runs without cache and it takes **2.655 sec**.
Second time, running with cache takes **1.190 sec**.

And it is possible to reuse the subquery for other queries.
```SQL
SELECT uniq(id)
FROM test_cache
WHERE (dt = '2019-12-05') AND (id IN 
(
    SELECT id
    FROM test_cache
    WHERE (dt >= '2019-12-01') AND (dt <= '2019-12-04')
    GROUP BY id
))
```
It also takes less time than without query-cache.

## Distributed Query Cache
Distributed Query Cache works on distributed queries.
It stores the result of remote-queries and reuses it later in the same queries.
If you want to use this feature, turn on an  option `SET use_experimental_distributed_query_cache = 1`.

**Create a test table**
```SQL
CREATE TABLE default.test_cache_dist
(
    `dt` Date, 
    `id` Int32, 
    `name` String
)
ENGINE = Distributed('test_cluster_two_shards', default, test_cache)
```

```SQL
DROP QUERY_CACHE;
```

```SQL
SET use_experimental_distributed_query_cache = 1;

SELECT count()
FROM test_cache_dist
WHERE (dt = '2019-12-05') AND (id IN 
(
    SELECT id
    FROM test_cache
    WHERE (dt >= '2019-12-01') AND (dt <= '2019-12-04')
    GROUP BY id
))
```

First time, this query runs without cache and it takes **3.932 sec**.
Second time, running with cache takes **0.003 sec**.

Notice that, the result of remote-queries are very large.
And it is not implemented to evict outdated distributed query results.
So that you have to carefully set this option.

## LRU
QueryCache evicts entries which are not used for a long time or are expired.
Cache starts to evict entries when their total weight exceeds `query_cache_memory_size` and when expiration time of these entries is due.

**Config.xml**
```XML
<yandex>
    <query_cache_memory_size>64000000000</query_cache_memory_size>
    <query_cache_min_lifetime>600</query_cache_min_lifetime>
</yandex>
```

## DROP QUERY_CACHE
If you want delete all the results in query-cache, use `drop query_cache`.
```SQL
DROP QUERY_CACHE
```

## Constraints
For now, it is only developed for MergeTree engines.

I worked on this feature at the previous Draft PR https://github.com/ClickHouse/ClickHouse/pull/7565.
And now I closed the Draft and created this PR.
Thank you.

### CODE REVIEW

**Comment by achimbab** [comment]
@alesapin 
I fixed some codes to pass the tests.
Let me know if there is any other things in order to start code review.
Thank you.

File: dbms/src/Interpreters/DatabaseAndTableWithAlias.h [hunk 1]
```diff
@@ -34,6 +35,18 @@ struct DatabaseAndTableWithAlias
...
+    bool operator<(const DatabaseAndTableWithAlias& rhs) const
+    {
+        if (database == rhs.database)
+            if (table == rhs.table)
+                return alias < rhs.alias;
+            else
+                return table < rhs.table;
+        else
+            return database < rhs.database;
+
```
Comment (azat) [inline dbms/src/Interpreters/DatabaseAndTableWithAlias.h]: `std::tie(database, table, alias) < std::tie(rhs.database, rhs.table, alias)` is simpler

Comment (achimbab) [reply dbms/src/Interpreters/DatabaseAndTableWithAlias.h]: @azat 
Thank you for your comment.

**Comment by achimbab** [comment]
@alexey-milovidov @alesapin @azat 
Hello~
I submitted this PR long time ago.
If any one review this PR, I will very appreciate.
Thank you.

**Comment by alexey-milovidov** [comment]
Sorry for the delay.
I will try to allocate the time today.

**Comment by alexey-milovidov** [comment]
I have created the PR with review changes: #8986
When it will be merged, your PR will also be merged automatically.

**Pull request closed without merging**